recon/src/recon.rs

```diff
@@ -429,23 +428,97 @@ where
     }
 }
 
+/// Represents reasons the store is unwilling to persist keys and values
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum InvalidItem<K: Key> {
+    /// The key or event data could not be parsed
+    InvalidFormat {
+        /// The key associated with the invalid data
+        key: K,
+    },
+    /// The data did not have a valid signature
+    InvalidSignature {
+        /// The key associated with the invalid data
+        key: K,
+    },
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+/// Represents items the store may be able to persist in the future
+pub struct PendingItem<K>
+where
+    K: Key,
+{
+    /// The key this item needs in order to be processed.
+    pub required_key: K,
+    /// The item that could not be stored. Can be retried in the future if
+    /// the `required_key` is discovered.
+    pub item: ReconItem<K>,
+}
+
+impl<K: Key> PendingItem<K> {
+    /// Construct a pending item
+    pub fn new(required_key: K, item: ReconItem<K>) -> Self {
+        Self { required_key, item }
+    }
+}
+
 /// The result of an insert operation.
-#[derive(Clone, Debug, PartialEq, Eq, Hash, Default)]
-pub struct InsertResult {
-    /// A true/false list indicating whether or not the key was new.
-    /// It is in the same order as the input list of keys.
-    pub keys: Vec<bool>,
+#[derive(Clone, Debug, PartialEq, Eq, Default)]
+pub struct InsertBatch<K>
+where
+    K: Key,
+{
+    /// The count of keys that were inserted in the batch.
+    new_cnt: usize,
+    /// The list of items that were considered invalid by the store.
+    /// Typically these events are "garbage" and the conversation should be ended or the
+    /// peer informed so they can clean up or stop sending the data.
+    pub invalid: Vec<InvalidItem<K>>,
+    /// Items that may be processed in the future but require more information to be interpreted.
+    /// For example, the store may need another item to interpret this event (e.g. `PendingItem::RequiresEvent`).
+    /// In the context of Ceramic, the init event is needed to verify if a data event is valid.
+    pub pending: Vec<PendingItem<K>>,
 }
 
-impl InsertResult {
+impl<K> InsertBatch<K>
+where
+    K: Key,
+{
     /// Construct an insert result
-    pub fn new(new_keys: Vec<bool>) -> Self {
-        Self { keys: new_keys }
+    pub fn new(new_cnt: usize) -> Self {
+        Self {
+            new_cnt,
+            invalid: Vec::new(),
+            pending: Vec::new(),
+        }
+    }
+
+    /// yah
+    pub fn item_count(&self) -> usize {
+        self.new_cnt + self.invalid.len() + self.pending.len()
+    }
+    /// yayayay
```

Comment:
oops, clean these doc comments up (or remove methods).. I added them when I was testing returning randomly selected items to mark pending and accidentally committed.